[MM-40821] contribute/server/schema-migration-guide: expand with new migration sys (#987)
mattermost/mattermost-developer-documentation: site/content/contribute/server/schema-migration-guide.md

Schema changes are always made synchronously when Mattermost starts up. This means the application won't be ready to serve requests until all schema changes are applied. In most cases, the new application won't be able to work until those schema changes are in place.

In a high availability environment, multiple instances will try to run migrations. To prevent that a lock table is used in the migration system therefore until migrations are completed none of the instances will start. Once the lock is released by a node, another instance will obtain the lock and it will be going to check the migrations table. Since the previous node already applied the migrations, other nodes are not going to apply migrations.

Since release 6.4 we have started to use a schema-based migration system. We are now creating SQL statement files to run migrations. A developer must create migration files for each database driver. Since we want our migrations to be reversible, the developer must create one `up` script along with a `down` script. For instance, a single migration would have the following files:

- `000066_upgrade_posts_v6.0.down.sql`
- `000066_upgrade_posts_v6.0.up.sql`

> The naming convention is the first part that will be used to determine the order in which the migrations should be applied and the next part until the `up|down.sql` suffix will be the migration name. We were using a database version before the new migration system hence the versions exist in the migration file name in the example. But using version identifiers for the next migration files is not mandatory. A developer can add any information to the name if they think it's going to be helpful.

Reviewer: ```suggestion
A file naming convention is used to determine the order in which the migrations should be applied that appends `up|down.sql` suffix to the migration name. We were using a database version before the new migration system which is why the versions exist in the migration file name in the example. Going forward, using version identifiers for future next migration files is not mandatory. A developer can add any information to the name if they think it's going to be helpful.
```

> We are using [morph](https://github.com/go-morph/morph) for the migration engine. The tool has a library and also a CLI. Mattermost server imports the library to have programmatic access to morph functions. A developer can use the morph CLI tool to test their migrations if they are working properly. Please follow instructions in the morph documentation about using the morph CLI tool.

Reviewer: ```suggestion
We are using [morph](https://github.com/go-morph/morph) for the migration engine. The tool has a library and also a CLI. Mattermost server imports the library to have programmatic access to morph functions. A developer can use the morph CLI tool to test whether their migrations are working properly. Please follow instructions in the morph documentation to use the morph CLI tool.
```

### I need to make a schema change. What do I do?

> 1. Add appropriate sql script file with the statements you want to run into the migrations directory. Which is located in `{project_dir}/db/migrations/{driver_name}/`. Do not forget to add script for `mysql` and `postgres`.

Reviewer: ```suggestion
1. Add the appropriate SQL script file containing the statements you want to run into the migrations directory. This directory is located in `{project_dir}/db/migrations/{driver_name}/`. Do not forget to add scripts for both `mysql` and `postgres` databases.
```

> 2. To embed the script run `make migrations-bindata`.

Reviewer: ```suggestion
2. To embed the script, run `make migrations-bindata`.
```

> 3. You can run the mattermost-server binary and the tool will automatically apply the migration if it is required. And the migration name will be saved into `db_migrations` table.

Reviewer: ```suggestion
3. When you run the mattermost-server binary, the tool will automatically apply the migration if it's required. The migration name will be saved in the `db_migrations` table.
```

### My migration has failed. What do I do?

> 1. If you think your migration is applied, and you want to revert changes: You can run the down script to rollback in a clean way. You can use morph CLI to apply down migrations.

Reviewer 2: An example would be good here for users to get an idea of what to do.

> 1. If you think your migration is applied, and you want to revert changes: You can run the down script to rollback in a clean way. You can use morph CLI to apply down migrations.

Reviewer: ```suggestion
1. If you think your migration is applied, and you want to revert changes, you can run the down script to roll back in a clean way. You can use morph CLI to apply down migrations.
```

> 2. The migration has been shipped in a release and you want to apply fixes: Instead of changing the existing script, you should add a new one so that `db_migrations` will stay consistent. You can edit the existing migration to be a no-op for future releases in this case.

Reviewer: ```suggestion
2. If the migration has been shipped in a release and you want to apply fixes, instead of changing the existing script, you should add a new one so that `db_migrations` will stay consistent. You can edit the existing migration to be a no-op for future releases in this case.
```
